### PR TITLE
Move Painless local methods to a dedicated FunctionTable

### PR DESCRIPTION
**Pull request by jdconrad**
This change moves LocalMethods out of Locals and into a separate FunctionTable which has the following advantages:
1) separation of responsibility between a set of global scope functions versus locally scoped variables
2) a dedicated class instead of a loose Map of functions to build both lambas and function pointers at run-time
3) it should be easier to re-build a separate variable table moving forward between the analysis pass and the writing pass during a future re-factoring
4) will be easier to manage with external passes during a future re-factoring
5) easier to understand/maintain

Note the majority of the change is mechanical since it was necessary to add the FunctionTable to the analysis method for each node. The major changes are to SSource, Locals, and the newly added FunctionTable.

### CODE REVIEW

**Comment by jdconrad** [comment]
@elasticmachine run elasticsearch-ci/bwc

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/SClass.java [hunk 1]
```diff
@@ -135,29 +136,30 @@ public void extractVariables(Set<String> variables) {
...
             function.generateSignature(painlessLookup);
 
-            String key = Locals.buildLocalMethodKey(function.name, function.parameters.size());
+            String key = FunctionTable.buildLocalFunctionKey(function.name, function.parameters.size());
 
-            if (methods.put(key,
-                    new LocalMethod(function.name, function.returnType, function.typeParameters, function.methodType)) != null) {
-                throw createError(new IllegalArgumentException("Duplicate functions with name [" + function.name + "]."));
+            if (table.getFunction(key) != null) {
+                throw createError(new IllegalArgumentException("Illegal duplicate functions [" + key + "]."));
```
Comment (rjernst) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/node/SClass.java]: nit: can this message be something like "X is already defined"?

**Review by rjernst — APPROVED** [review]
LGTM

**Comment by jdconrad** [comment]
@elasticmachine run elasticsearch-ci/bwc

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/SClass.java (hunk 1 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/node/SClass.java]: Fixed.

**Comment by jdconrad** [comment]
@rjernst Thanks for the review! Will commit as soon as CI passes.

**Comment by jdconrad** [comment]
@elasticmachine test this please

**Comment by jdconrad** [comment]
@elasticmachine run elasticsearch-ci/1

**Comment by jdconrad** [comment]
@elasticmachine run elasticsearch-ci/1

**Pull request merged**